TASK: Render menu prototypes without fluid

mficzel opened the pull request:
The prototypes `Neos.Neos:Menu`, `Neos.Neos:DimensionsMenu` and `Neos.Neos:BreadcrumbMenu` are now implemented in pure fusion. The prototype `Neos.Neos:MenuItemListRenderer` is added that is used by those prototypes to render the result as a very basic unordered list.

The option `calculateItemStates` is added to the prototypes `Neos.Neos:MenuItems`, `Neos.Neos:DimensionsMenuItems` and `Neos.Neos:BreadcrumbMenuItems` with the default value being false. Once enabled the itemState of each menuItem is calculated. While the calculation of itemState is not overly expensive the calculation depends on the current documentNode and this is only valid for the single document. This makes sharing menu caches between documents harder.

In addition the following issues are fixed:
- `DimensionMenuItems` did not contain a url
- `MenuItems` did not calculate item states other than normal
-  the signature of `NodeAggregateIds::fromNodes(array $nodes)` is changed to `NodeAggregateIds::fromNodes(Nodes $nodes)` and the method
`NodeAggregateIds::contain(NodeAggregateId $nodeAggregateId)` is added. There were no users of this to be found so this should be safe.
- `RenderAttributesTrait` will now use the value of backedEnums which allows to use Enums in `Neos.Fusion:Tag.attributes` and `Neos.Fusion:Augmenter` this was needed for the `MenuItemListRenderer`

**Upgrade instructions**

Since the Menus are no longer Templates the `templatePathAndFilename` was removed together with all state specific attribute rendering properties. You should adjust your rendering accordingly or create custom Menu prototypes that either have a renderer the suits your needs or are based on a Fluid Template again.

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mficzel committed:
TASK: Render menu prototypes without fluid

The prototypes `Neos.Neos:Menu`, `Neos.Neos:DimensionsMenu` and `Neos.Neos:BreadcrumbMenu` are now implemented in pure fusion.
The prototype `Neos.Neos:MenuItemListRenderer` is added that is used by those prototypes to render the result as a very basic unordered list.

The flag `calculateItemStates` is added to the prototypes `Neos.Neos:MenuItems`, `Neos.Neos:DimensionsMenuItems` and `Neos.Neos:BreadcrumbMenuItems` with the default value being false. Once enabled the itemState of each menuItem is calculated.

In addition the following issues are fixed:
- DimensionMenuItems did not contain a url
- MenuItems did not calculate item states.

ahaeslich commented on Neos.Neos/Classes/Fusion/MenuItem.php:
```diff
@@ -11,62 +11,22 @@
 
     /**
      * @return array<int,MenuItem>
+     * @deprecated Use getChildren instead
```
Is this correct? Now `getSubItems` and `getChildren` are deprecated 🤔 .

mficzel replied:
Good catch copy paste error. Still want to improve this as the calculation of items could benefit a lot from using more recent cr methods

mficzel replied:
adjusted

mficzel commented:
Have a question wether we really need the calculateItemState flag if the cost is merely a single `findAncestors`. 

Also the findAncestors could be used for detecting the starting point aswell so we ideally would end up with 2 queries  `findAncestors` and `findSubtree` for the generation of a menu.

ahaeslich commented on Neos.Neos/Resources/Private/Fusion/Prototypes/MenuItemListRenderer.fusion:
```diff
@@ -0,0 +1,17 @@
+    <Neos.Fusion:Loop items={props.items} itemName="item">
+      <li class={item.state}>
+        <Neos.Fusion:Tag tagName={item.uri ? 'a' : 'span'} attributes.href={item.uri} attributes.title={item.label}>{item.label}</Neos.Fusion:Tag>
+        <Neos.Neos:MenuItemListRenderer @if={item.subItems} items={item.subItems} />
```
And will we still use `subItems` here or `children`?

mficzel replied:
adjusted

mhsdesign commented:
> Have a question wether we really need the calculateItemState flag if the cost is merely a single findAncestors.

Good point indeed. Hmm we seem to need the ancestors to highlight the trace:

<img width="615" alt="image" src="https://github.com/neos/neos-development-collection/assets/85400359/e9a76dec-f8a2-4782-a9de-8f54d6a392cc">


But when we would use `findSubtree` as you suggested, which we dont do at the moment? The ancestor information would ship free house as its the parent in the tree ... in that case its no expense at all to calculate and we can drop the flag without doubt.

But i think that @bwaidelich is currently creating test cases for the menu so that we can refactor it to use `findSubtree` - so thats on the list for a next pr.

mficzel commented:
> But i think that @bwaidelich is currently creating test cases for the menu so that we can refactor it to use findSubtree - so thats on the list for a next pr.

@mhsdesign that is a good point and could help to keep this pr in scope. I thought about refactoring this now but it makes more sense with tests.

Regarding the `calculateItemState`  i think it makes sense to have and to disable it by default because those states make sharing of menu caches between documents more "interesting".

mficzel commented:
> But when we would use findSubtree as you suggested, which we dont do at the moment? The ancestor information would ship free house as its the parent in the tree ... in that case its no expense at all to calculate and we can drop the flag without doubt.

I think first should be a findAncestors from currentDocument up. Needed to determine the starting point via entryLevel. From there a single findSubtree should be enough. 2 queries for a menu sounds quite ok to me.

Sebobo commented on Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php:
```diff
@@ -70,6 +77,20 @@ abstract class AbstractMenuItemsImplementation extends AbstractFusionObject
     protected ContentRepositoryRegistry $contentRepositoryRegistry;
 
+    /**
+     * Should nodes that have "hiddenInIndex" set still be visible in this menu.
```
Looks like a copy&paste comment ;)

mhsdesign replied:
I proposed a removal here: https://github.com/neos/neos-development-collection/pull/4623/files/8e04b9ad9f83857c6c654d02f79e3add20c4a1b4#r1372881520

bwaidelich commented on Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php:
```diff
@@ -60,6 +60,13 @@ abstract class AbstractMenuItemsImplementation extends AbstractFusionObject
 
+    /**
+     * Internal cache for the calculateItemStates property.
+     *
+     * @var boolean
+     */
+    protected $calculateItemStates;
```
nitpick, a little more defensive:
```suggestion
     */
    private ?bool $calculateItemStates = null;
```

mficzel replied:
Why not directly 
```php
private bool $calculateItemStates = false;
```
Unsure abut adding this single property private when there are lots of protected ones around.

bwaidelich replied:
because you just use this field as runtime cache for `isCalculateItemStatesEnabled()` and use `null` to distinguish the cached from the uncached case, remember? :)
I started suffixing those fields with some `...RuntimeCache` to make that clearer.

I'd say it's OK to introduce private fields even if we still use `protected`, especially if it is just an internal field with a dedicated accessor (like here). But it was just a nitpick anyways

mhsdesign replied:
yes its more of a drop of water on a hot stone? I would refrain from discussing such minor things 😅

bwaidelich commented on Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php:
```diff
@@ -70,6 +77,20 @@ abstract class AbstractMenuItemsImplementation extends AbstractFusionObject
+    /**
+     * Should nodes that have "hiddenInIndex" set still be visible in this menu.
+     *
+     * @return boolean
```
```suggestion
     * Whether the active/current state of menu items is calculated on the server side.
     * This has an effect on performance and caching
```

bwaidelich commented on Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php:
```diff
@@ -169,4 +190,15 @@ protected function getCurrentNodeRootline(): array
         return $this->currentNodeRootline;
     }
+
+    protected function buildUri(Node $node): string
```
thought: IMO we should stop using `protected` most of the time, but that's out of scope I suppose

mhsdesign replied:
yes the php integration of fusion is more of an super ugly layer - i wouldnt go there if you want to see nice looking php - id be fine with keeping it as is

mficzel replied:
Only Moved this up from the MenuImplementation as DimensionMenu also needs URLs. Would abstain from further changes.

bwaidelich commented on Neos.Neos/Classes/Fusion/DimensionMenuItem.php:
```diff
@@ -0,0 +1,25 @@
+        public readonly ?MenuItemState $state = null,
+        public readonly ?string $label = null,
+        public readonly ?array $targetDimensions = null,
+        public readonly ?string $uri = null
```
nitpick: The doc comment could use some love and the class could be `readonly`
```suggestion
 * A menu item for dimension menus
 * Compared to the default {@see MenuItem} it has no `menuLevel` property, but one for the `targetDimensions`
 */
final readonly class DimensionMenuItem
{
    /**
     * @param array<string,mixed>|null $targetDimensions
     */
    public function __construct(
        public ?Node $node,
        public ?MenuItemState $state = null,
        public ?string $label = null,
        public ?array $targetDimensions = null,
        public ?string $uri = null
```

also, could we add some doc comments about the potential values of `$targetDimensions` – is that really an array or could we use one of the existing VOs?

bwaidelich commented on Neos.Neos/Classes/Fusion/MenuItem.php:
```diff
@@ -11,58 +11,17 @@
+        public readonly ?string $label = null,
+        public readonly int $menuLevel = 1,
+        public readonly array $children = [],
+        public readonly ?string $uri = null
```
nitpick: Like above I would suggest to make the class `readonly` (can't suggest that because it's not part of the changeset)
```suggestion
        public Node $node,
        public ?MenuItemState $state = null,
        public ?string $label = null,
        public int $menuLevel = 1,
        public array $children = [],
        public ?string $uri = null
```

mficzel replied:
Adjusted as suggested

bwaidelich commented on Neos.Neos/Classes/Fusion/MenuItemsImplementation.php:
```diff
@@ -312,14 +318,32 @@ protected function traverseUpUntilCondition(Node $node, \Closure $callback): voi
+            return $this->currentNodeAncestors;
+        }
+        $subgraph = $this->contentRepositoryRegistry->subgraphForNode($this->currentNode);
+        $this->currentNodeAncestors = $subgraph->findAncestorNodes(
```
This will potentially load thousands and thousands of nodes into memory.. I don't think that we can pre-cache this

mhsdesign replied:
it should be super fine as this is opt in behavior.

bwaidelich replied:
Really?
Do you think an integrator is aware that we load all pages into memory if they set this flag to true to get the Neos < 9 behavior ??

mhsdesign replied:
No, but my argument would be: If you want speed you know that you only want one cache entry https://mind-the-seb.de/blog/globally-cached-fragments-in-neos-cms and then you know that you dont want to have this calculate item flag -> speeeeeed.

bwaidelich replied:
I think we are talking cross purpose.
I agree of having that flag for exactly that reasons.
I don't agree of a default implementation that loads all documents into memory – especially because you will only see a crazy increase of RAM usage and no direct indication to why that is

mficzel replied:
Not sure i understand that correctly. As far as i see `findAncestorNodes` leads to a single (internally recursive) db-query that will return only the nodes up from the given node->id. 1000s of nodes would require a structure 1000 levels deep that will cause other issued aswell. 

Is that somehow correct or am i on the woodway.

bwaidelich replied:
You're right.. I confused myself and mixed up descendants and ancestors for a moment #coffee 😶‍🌫️

bwaidelich replied:
Sorry for the confusion. Loading the _ancestors_ once per menu should be fine.
But I would still suggest to only store the node *ids*  instead of the instances with all their properties etc.

We can use `NodeAggregateIds` VO for that (and should extend that with some `contains(NodeAggregateId $nodeAggregateId): bool` to simplify the check below)

mficzel replied:
Adjusted as requested

bwaidelich reviewed (commented):
Great work, this looks and feels so much nicer already! <3

A couple of thoughts (in addition to the - mostly nitpicky – comments):

We should always have large sites in the back of our heads. Even if we only load document ancestor nodes, those can be tens of thousands so we should not load them into memory (Neos already has quite an excessive RAM usage).

So calculating the `active` state will probably have to be calculated for every item in the menu.

But IMO the bigger issue is that this results in an individual cache entry for every menu for every page which – in turn – means that you'll basically have to flush the whole content cache if you change anything in the menu implementation.

That's why I like the `calculateItemState` prop and IMO we could even make the `@cache` configuration depend on it, such that we create a common cache entry if it is disabled

mficzel commented:
Short note. After this is merged and with tests for menu items in place we should try to cleanup the dependencies between the MenuItems prototypes (that all are derived from MenuItems). 

I would like all *MenuItems prototypes to be standalone fusion prototypes with an implementation class.

bwaidelich commented:
>  After this is merged and with tests for menu items in place we should try to cleanup the dependencies between the MenuItems prototypes

Yes, great idea! And we could use that chance to replace the `MenuItemState` class with a proper enum (and potentially remove the `absent` case that should IMO never happen)

mficzel commented:
Absent is a state that can happen in dimension menus. Using enums is a good idea we can aswell implement here.

mficzel committed:
TASK: Store NodeAggregateIds of current node rootline instead of full nodes

This is more memory efficient and we only compare the AggregateIds anyways

In addition the signature of `NodeAggregateIds::fromNodes(array $nodes)` is changed to `NodeAggregateIds::fromNodes(Nodes $nodes)` and the method
`NodeAggregateIds::contain(NodeAggregateId $nodeAggregateId)` is added. There were no users of this to be found so this should be safe.

mficzel committed:
Update Neos.Neos/Classes/Fusion/DimensionMenuItem.php

Co-authored-by: Bastian Waidelich <b.waidelich@wwwision.de>

mficzel committed:
Update Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php

Co-authored-by: Bastian Waidelich <b.waidelich@wwwision.de>

mficzel committed:
TASK: Make MenuItem and DimensionMenuItemReadonly and convert MenuItemState to a BackedEnum

In addition the RenderAttributesTrait was adjusted to handle backed enums like stringables.

mficzel commented:
@bwaidelich i think your issues are addressed. We could think about backporting the backedEnum in attributes support to 8.3 but i would decide upon that later.

bwaidelich commented:
> i think your issues are addressed

@mficzel you're the man, this looks awesome!

Just one last thing for now (and we can also postpone it to a separate PR): Do you think it makes sense to make the `@cache` config dependent on the `calculateItemState` flag such that the same cache entry can be shared?

mficzel commented:
@bwaidelich not safely as entryLevel -x also depends on the current document

would try that separately

bwaidelich reviewed (approved):
Love it! +1 by reading

mhsdesign commented on Neos.Neos/Resources/Private/Fusion/Prototypes/Menu.fusion:
```diff
@@ -1,48 +1,18 @@
-    filter = ${this.filter}
-    renderHiddenInIndex = ${this.renderHiddenInIndex}
-    itemCollection = ${this.itemCollection}
+  @private {
+    items = Neos.Neos:MenuItems {
+      @ignoreProperties = ${['attributes']}
+      @apply.props = ${props}
+    }
```
Hmm this `@apply.props = ${props}` is quite a clever pattern. But im torn apart whether to use it here. 

We lately put some work into it to define for every prototype its defaults https://github.com/neos/neos-development-collection/pull/3915.
And since the amount of properties for the menu api are final i like to propose to do here the same (again).

mficzel replied:
Yeah. Would do that aswell but during the untangling of the menu prototype inheritance. 

So in a later pr. We will have to check which property really applies to each type of menu then. With the current inheritance we would have to cement the mess further.

mhsdesign replied:
Okay sound like a good plan ;)

Then let’s do it later 👍

mhsdesign commented:
I like to wait with merges until Bastians tests are ready ;)

mficzel commented:
@mhsdesign this changes nothing that was not broken/missing before. Other than the renderers only the state calculation and uris for the dimension menu were added again. By waiting we will only make this more complex by fixing more stuff in a single pr.

bwaidelich commented:
> I like to wait with merges until Bastians tests are ready ;)

For me it's the other way around.
I would suggest to merge this one ASAP so that I can rebase the test-PR on it.
The tests will fail because `MenuItemsImplementation` still behaves differently than < Neos 9. But we can then start to rewrite it (replacing nasty array juggling with Subgraph queries along the way) in that PR until the tests are green

bwaidelich reviewed (approved):
There's potentially some follow up work to be done (backports, more tests, cache mode tricks to allow reusability) but this can all be done in separate PRs.

I really love this, thanks for taking care! <3

Sebobo commented on Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeAggregateIds.php:
```diff
@@ -92,6 +90,11 @@ public function merge(self $other): self
         ));
     }
 
+    public function contain(NodeAggregateId $nodeAggregateId): bool
```
Shouldn't it be called `contains`?
```suggestion
    public function contains(NodeAggregateId $nodeAggregateId): bool
```

bwaidelich replied:
I was wondering too, but `NodeAggregateIds` is a plural, so "contain" is gramatically correct I think

mhsdesign replied:
yey singular makes sense in that case ^^. `$nodeAggregateIds->contain($id)`

Alternative would be `$nodeAggregateIds->has($id)`

Sebobo replied:
True

Sebobo commented on Neos.Neos/Resources/Private/Fusion/Prototypes/DimensionsMenu.fusion:
```diff
@@ -1,34 +1,29 @@
+      documentNode = ${Neos.Caching.entryIdentifierForNode(documentNode)}
+    }
+    entryTags {
+      1 = ${Neos.Caching.nodeTypeTag('Neos.Neos:Document', documentNode)}
```
Why do we flush on any document change and not just the current one?

mficzel replied:
Because any document below the starting point could require invalidation.

So descendantOfTag with the startingPoint would be the right choice but that collides with this entry level magic. Also it would invalidate on any content change below.

I really think we should make startingPoint a requirement with default ${site} but that is for another pr.

kitsunet reviewed (approved):
Thank you! Amazing :)

Sebobo commented on Neos.Neos/Resources/Private/Fusion/Prototypes/BreadcrumbMenu.fusion:
```diff
@@ -1,17 +1,29 @@
+      documentNode = ${Neos.Caching.entryIdentifierForNode(documentNode)}
+    }
+    entryTags {
+      1 = ${Neos.Caching.nodeTypeTag('Neos.Neos:Document', documentNode)}
```
Could we only include the document nodes parents for flushing?

mficzel replied:
We already flush for all documents

Sebobo replied:
So this way the DimensionsMenu and the Breadcrumb of every site is flushed when any document is changed in the website...

So IMO just including parents and the current node for both cases would solve both and reduce the flushing by a lot

mficzel replied:
Decided not to do this as this pr is about de-fluidizing not optimizing. Before the prototype inherited the cache annotations of menu which flush for all documents.

There are plans to optimize the menus further so this is only the first step. This pr should not make anything worse but as it is but also should not need to fix more than it promises.

Would like to get this merged to look into the optimizing topic without piling too many changes upon each other.

mhsdesign reviewed (approved):
> I would suggest to merge this one ASAP so that I can rebase the test-PR on it.

yes i agree as discussed ;)

mhsdesign commented on Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php:
```diff
@@ -70,6 +77,20 @@ abstract class AbstractMenuItemsImplementation extends AbstractFusionObject
+     * Should nodes that have "hiddenInIndex" set still be visible in this menu.
+     *
+     * @return boolean
+     */
```
```suggestion
```

Sebobo reviewed (approved):
Fine for now. Just hope the caching topic for this case doesn't get lost. Performance is really important and didn't get enough attention in the past, especially if we want to intrigue our users with the performance of the new CR.

bwaidelich commented:
> Just hope the caching topic for this case doesn't get lost

I agree!
We focused on a 1:1 replacement for now, but the new implementation already does not calculate the item state per menu which will allow for a single reusable cache entry
